daft/execution/physical_plan.py

```diff
@@ -396,6 +399,628 @@ def broadcast_join(
+                    larger_requests if left_is_larger else smaller_requests,
+                    smaller_requests if left_is_larger else larger_requests,
+                )
+                yield None
+            # If we just exhausted small side of join, set smaller done flag.
+            elif yield_smaller and not smaller_done:
+                smaller_done = True
+            # If we just exhausted larger side of join, set larger done flag.
+            elif not yield_smaller and not larger_done:
+                larger_done = True
+            # We might still be waiting for some merge-join tasks to complete whose output we still need
+            # to coalesce.
+            elif not merge_join_task_tracker.all_tasks_done():
+                logger.debug(
+                    "merge join blocked on completion of merge join tasks (pre-coalesce).\nMerge-join tasks: %s",
+                    list(merge_join_task_tracker._finalized_tasks.values()),
+                )
+                yield None
+            # Otherwise, all join inputs are done and all merge-join tasks are done, so we are entirely done emitting
+            # merge join work.
+            else:
+                return
+
+
+def _is_strictly_bounded_above_by(
+    lower_part: SingleOutputPartitionTask[PartitionT], upper_part: SingleOutputPartitionTask[PartitionT]
+) -> bool:
+    """
+    Returns whether lower_part is strictly bounded above by upper part; i.e., whether lower_part's upper bound is
+    strictly less than upper_part's upper bound.
+    """
+    lower_boundaries = lower_part.partition_metadata().boundaries
+    upper_boundaries = upper_part.partition_metadata().boundaries
+    assert lower_boundaries is not None and upper_boundaries is not None
+    return lower_boundaries.is_strictly_bounded_above_by(upper_boundaries)
+
+
+def _memory_bytes_for_coalesce(input_parts: Iterable[SingleOutputPartitionTask[PartitionT]]) -> int | None:
+    # Calculate memory request for task.
+    size_bytes_per_task = [task.partition_metadata().size_bytes for task in input_parts]
+    non_null_size_bytes_per_task = [size for size in size_bytes_per_task if size is not None]
+    non_null_size_bytes = sum(non_null_size_bytes_per_task)
+    if len(size_bytes_per_task) == len(non_null_size_bytes_per_task):
+        # If all task size bytes are non-null, directly use the non-null size bytes sum.
+        size_bytes = non_null_size_bytes
+    elif non_null_size_bytes_per_task:
+        # If some are null, calculate the non-null mean and assume that null task size bytes
+        # have that size.
+        mean_size = math.ceil(non_null_size_bytes / len(non_null_size_bytes_per_task))
+        size_bytes = non_null_size_bytes + mean_size * (len(size_bytes_per_task) - len(non_null_size_bytes_per_task))
+    else:
+        # If all null, set to null.
+        size_bytes = None
+    return size_bytes
+
+
+def sort_merge_join(
```

Comment:
Spoke Offline, favor inserting sorts in physical plan rather than doing a combined sort_merge_join

Reply:
Added an execution config flag for this, defaults to inserting sorts, but we can still turn on the old aligned boundaries sort path in future benchmarking.